shelley-ma/formal-spec/value.tex

```diff
@@ -80,46 +228,111 @@ \subsection*{Representing Multi-asset Types and Values}
+
+    If $(\var{pid}, \var{aname})$ is an $\AssetID$ and $v \in \Value$,
     the quantity in $v$ belonging to that $\AssetID$ is $v~\var{pid}~\var{aname}$.
-    Tokens are fungible with each other if and only if they belong to the same $\AssetID$,
-    i.e. they have the same $\PolicyID$ and $\AssetName$. Terms of type $\Value$ are sometimes also referred to as
+    Terms of type $\Value$ are sometimes also referred to as
     \emph{token bundles}.
 
-  \item $\mathsf{adaPolicy}$ and $\mathsf{adaName}$ are the $\PolicyID$ and $\AssetName$ of Ada respectively.
-    It is not possible to mint any token with the $\PolicyID$ $\mathsf{adaPolicy}$, which means that every
-    token in the UTxO belonging to $\mathsf{adaPolicy}$ will have $\AssetName$ $\mathsf{adaName}$.
+   To be precise, the multi-asset
+   type, $\Value$ is a finitely supported function of two variables,
+   $\PolicyID$ and $\AssetName$ to $\Quantity$.
 
-  \item $\mathsf{size}$ returns the size (in units of memory, eg. bytes or words) of a $\Value$.
-  We give this function, alongside implementation-specific constants we have used in it in Section \ref{sec:value-size}.
+  \item $\fun{inject}$ takes a $\Coin$ amount and returns a $\Value$ containing
+  that amount of Ada and no other assets.
 
-  \item $\fun{inject}$ and $\fun{coin}$ convert between the two types,
-  by building a $\Value$ that only contains Ada given by the input, and extracting
-  the amount of Ada as (a $\Coin$) from a $\Value$ respectively.
-  Note that composed as
+  \item $\fun{coin}$ returns the amount of Ada contained in a $\Value$ term (ie.
+  the first term of the pair).
+\end{itemize}
 
-  \[\fun{coin} \circ \fun{inject} = \fun{id}_{\Coin}\]
+\textbf{The Ada special case.} Making Ada a special case in the $\Value$ definition is useful
+because Ada is the most common type of asset on the ledger. It also can
+be used for special purposes (such as fees) for which other assets cannot be used.
+Moreover, it is unlikely that tokens other than Ada can ever be used for this
+purpose, as the underlying consensus algorithm relies on Ada in a way that
+cannot be extended to user-defined tokens.
+We sometimes refer to Ada as the primary or principal currency. Ada does not,
+for the purposes of the Mary ledger spec, have a $\PolicyID$ or an $\AssetName$.
 
-  these functions give the identity on $\Coin$, but composed in the opposite order as
+\subsection{Value Operations and Partial Order}
+To fully describe $\ValClass$ instantiation with $\Value$, we require
+additional operations on $\Value$, including binary relations, addition, and
+scalar multiplication. These are all done
+pointwise in accordance with the usual definitions of these operations on
+finitely supported maps, see Figure \ref{fig:pointwise}.
 
-  \[\fun{inject} \circ \fun{coin}\]
+\begin{figure*}[t!]
+  \begin{align*}
+    & (c, v)~+~(d, w) =& (c+d, ~\{~ pid\mapsto_0 (aname\mapsto_0 (v~\var{pid}~\var{aname})~+~(w~\var{pid}~\var{aname}) ~\vert~ \\
```

Comment:
This is really difficult to read (in the PDF). This would also get easier if `Value` gets changed back to the old type

Reply:
It is hard to read, but i think it does not get us anything to use a type that's almost the real type used in the code, it only makes comparing the two more confusing

Reply:
I think we discussed the definition of `Value` a lot in the past, and we settled on the old definition for the spec, and the optimized definition for the code. The main benefit of abstracting it like with this PR is that we can actually mention both in the spec. If we only wanted the code and the spec to agree, we could avoid all that and just could have changed the definition of `Value`. I think what we really should be doing is keeping the old definition of `Value`, spell out the isomorphism between it and the optimized version in the implementation and then say something along the lines of: all the functions for the optimized `Value` type of the implementation are derived via transporting them along that isomorphism.